ompi/mca/hook/base/hook_base.c

```diff
@@ -0,0 +1,299 @@
+
+    additional_callback_components = OBJ_NEW(opal_list_t);
+
+    /* Open up all available components */
+    ret = mca_base_framework_components_open( &ompi_hook_base_framework, flags );
+    if (ret != OMPI_SUCCESS) {
+        return ret;
+    }
+
+    /*
+     * Make sure that the `MCA_BASE_COMPONENT_FLAG_REQUIRED` components defined
+     * as static are loaded. If we find one that was avoided then error out.
+     */
+    if( NULL != static_components ) {
+        for (int i = 0 ; NULL != static_components[i]; ++i) {
+            if( static_components[i]->mca_component_flags & MCA_BASE_COMPONENT_FLAG_REQUIRED ) {
+                // Make sure that this component is in the list of components that
+                // were included in the earlier framework_components_open() call.
+                found = false;
+                OPAL_LIST_FOREACH(cli, &ompi_hook_base_framework.framework_components, mca_base_component_list_item_t) {
+                    component = (mca_base_component_t*)cli->cli_component;
+                    if( component == static_components[i] ) {
+                        found = true;
+                        break;
+                    }
+                }
+                if( !found ) {
+                    opal_show_help("help-mca-hook-base.txt", "hook:missing-required-component", true,
+                                   ompi_hook_base_framework.framework_name,
+                                   static_components[i]->mca_component_name);
+                    return OPAL_ERR_NOT_SUPPORTED;
+                }
+            }
+        }
+    }
+
+    /* Assume that if the component is present then it wants to be used.
+     * It has the option to have NULL as the function pointer for any
+     * functions call hook locations they do not want to hear about
+     */
+
+    /*
+     * There are three classes of components - we want them processed in this order
+     * 1) static components
+     * 2) dynamic components
+     * 3) internal 'component' hooks (from other places in the code)
+     *
+     * The ordering of (1) and (2) is managed by mca_base_component_find().
+     * We keep a separate list for the 'internal' hooks.
+     */
+
+    ompi_hook_is_framework_open = true;
+
+    /* All done */
+    return OMPI_SUCCESS;
+}
+
+static int ompi_hook_base_close( void )
+{
+    int ret;
+
+    /*
+     * Close our components
+     */
+    ret = mca_base_framework_components_close( &ompi_hook_base_framework, NULL );
+    if( OMPI_SUCCESS != ret ) {
+        return ret;
+    }
+    ompi_hook_is_framework_open = false;
+
+    return OMPI_SUCCESS;
+}
+
+
+int ompi_hook_base_register_callbacks(ompi_hook_base_component_t *comp)
+{
+    mca_base_component_list_item_t *cli;
+    bool found = false;
+
+    // Check if it is already there
+    OPAL_LIST_FOREACH(cli, additional_callback_components, mca_base_component_list_item_t) {
+        if( cli->cli_component == (mca_base_component_t*)comp ) {
+            found = true;
```

Comment:
To-MAY-to, to-MAH-to: you could just `return OMPI_SUCCESS` here and skip the use of `found` altogether.  Doesn't really matter either way -- just makes the code 0.00003% simpler.  😄